### PR TITLE
docs: Add BuildOnFedora.md

### PR DESCRIPTION
**Pull request by soonho-tri**
This change is [Reviewable](https://reviewable.io/reviews/dreal/dreal4/227)

### CODE REVIEW

**Comment by soonho-tri** [comment]
@DKPeterson, can you test this?

**Comment by DKPeterson** [comment]
Hi, SoonHo—

Thank you for the quick response.

I have to let you know that I’m a n00b at this, especially at using git. So rather than bluff my way though this—and likely misleading you in the process—I think it’s better if I ask you exactly what you want me to do.

I think you’re asking me to pull from a new branch from the repo, and then re-run my bazel build of dreal. But you need to specify exactly what commands I should use; if you leave it to me to bushwhack it’ll confuse us both.

Here’s what my git repo looks like:

$ git branch -a
* master
  remotes/origin/HEAD -> origin/master
  remotes/origin/gh-pages
  remotes/origin/master

What do you want me to do from here?

====================================================

I have found out a couple of important things since my first post:

Initially, the error I was getting came from bazel’s attempt to use pkg-config on ibex:

ERROR: An error occurred during the fetch of repository 'ibex’:
       Unable to complete pkg-config setup for @ibex repository: error 1 from [/usr/bin/pkg-config, "ibex"]: 

After googling for a fix, I tried specifying the PKG_CONFIG_PATH environment variable (by default it’s empty, my ibex.pc file is in /usr/local/share/pkgconfig/. Same problem for the nlopt.pc file, whose .pc file is in /usr/local/lib64/pkgconfig/). So I did this:

export PKG_CONFIG_PATH=/usr/local/share/pkgconfig/:/usr/local/lib64/pkgconfig/:$PKG_CONFIG_PATH

It did not work. That is, bazel’s attempt to locate these packages (using pkg-config) still failed.

Next, I decided to add the correct RHEL7 pkgconfig paths (or, at least, the ones on my VM, after installing ibex and nlopt) by directly editing the .../dreal4/dreal/workspace.bzl code. This entailed adding two lines to the definition of dreal_workspace(), one for ibex and another for nlopt:

       "/usr/local/share/pkgconfig”  # for ibex
       "/usr/local/lib64/pkgconfig/“ # for nlopt

This worked! bazel was now able to find the .pc info for these packages.

The next problem was with gcc. I got this error:

ERROR: /home/kent/.cache/bazel/_bazel_kent/223653a5c28b53068ec3c0801e54a51c/external/com_google_absl/absl/hash/BUILD.bazel:27:1: C++ compilation of rule '@com_google_absl//absl/hash:hash' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 35 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
gcc: error: unrecognized command line option '-std=c++14'

To fix this, I used an apparently common RedHat developer trick. The default developer environment has an old gcc (v.4.8.5). It is possible, using RHEL's Software Collections (scl) directive to temporarily swap out (i.e., up/down-grade gcc, for instance) devtoolsets. So I did this:

scl enable devtoolset-7 ‘bash’

which now gives me a newer version of gcc (v.7.3.1) in my local shell. Now, when I run bazel //… , it gets past the gcc error.

But it now throws this loader error:

ERROR: /home/kent/sally/packages/dreal4/test/BUILD.bazel:4:1: Linking of rule '//test:abseil_test' failed (Exit 1) gcc failed: error executing command /opt/rh/devtoolset-7/root/usr/bin/gcc @bazel-out/k8-opt/bin/test/abseil_test-2.params

Use --sandbox_debug to see verbose messages from the sandbox
bazel-out/k8-opt/bin/_solib_k8/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Stypes_Slibbad_Uvariant_Uaccess.so: error: undefined reference to 'operator delete(void*, unsigned long)'
bazel-out/k8-opt/bin/_solib_k8/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Stime_Sinternal_Scctz_Slibtime_Uzone.so: error: undefined reference to 'std::__throw_out_of_range_fmt(char const*, …)'
.
.   <eight more “undefined reference to...” errors, just like these>
.
bazel-out/k8-opt/bin/_solib_k8/libexternal_Scom_Ugoogle_Uabsl_Sabsl_Sbase_Slibthrow_Udelegate.so: error: undefined reference to 'std::underflow_error::underflow_error(char const*)'
collect2: error: ld returned 1 exit status


That’s where I am now. I haven’t found any useful google hits for fixing this problem.

—Kent

**Comment by soonho-tri** [comment]
Hi,

1) I have tested it on a VM running Fedora 32. I'll merge it so that you don't need a special trick to check this PR out.

2) I have no experience regarding RHEL/Fedora actually. Having said that, the key part here is to build and install ibex-lib which is a core dependency of dReal. I hope the instructions above are good enough to reproduce in your system.

3) The gcc error looks similar to https://github.com/google/verible/issues/78.

**Comment by soonho-tri** [comment]
PTAL https://github.com/dreal/dreal4/blob/6f39e530024f536d26161a07fe0774caa544789d/docs/BuildOnFedora.md#how-to-build-and-install-dreal-on-fedora-32

**Comment by DKPeterson** [comment]
TL;DR  Do I have to build ibex from “ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca”? If so, then there is a problem, because waf does not recognize the specified options.

====================================================

From your latest instructions, I notice that bazel v.3+ is needed (this is not evident from this <https://docs.bazel.build/versions/3.1.0/install-redhat.html> bazel web page).

When I build with bazel-3.4.1 (and libstdc++.a  and  gcc 7.3.1), I get this error:

$ BAZEL_LINKLIBS=-l%:libstdc++.a bazel build //dreal:dreal
Starting local Bazel server and connecting to it...
... still trying to connect to local Bazel server after 10 seconds ...
INFO: Analyzed target //dreal:dreal (42 packages loaded, 1691 targets configured).
INFO: Found 1 target...ERROR: /home/kent/sally/packages/dreal4/dreal/contractor/BUILD.bazel:14:17: C++ compilation of rule '//dreal/contractor:contractor' failed (Exit 1) gcc failed: error executing command /opt/rh/devtoolset-7/root/usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG ... (remaining 92 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
In file included from dreal/contractor/contractor_cell.cc:8:0:
./dreal/contractor/contractor_ibex_polytope.h:63:25: error: 'LinearizerCombo' is not a member of 'ibex'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
./dreal/contractor/contractor_ibex_polytope.h:63:25: note: suggested alternative: 'LinearizerCompo'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
                         LinearizerCompo
./dreal/contractor/contractor_ibex_polytope.h:63:25: error: 'LinearizerCombo' is not a member of 'ibex'
./dreal/contractor/contractor_ibex_polytope.h:63:25: note: suggested alternative: 'LinearizerCompo'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
                         LinearizerCompo
./dreal/contractor/contractor_ibex_polytope.h:63:40: error: template argument 1 is invalid
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                                        ^
./dreal/contractor/contractor_ibex_polytope.h:63:40: error: template argument 2 is invalid
Target //dreal:dreal failed to build
Use --verbose_failures to see the command lines of failed build steps.

It looks like the ibex library I have (v.2.8.7) does not have the LinearizerCombo member. Looks like there might be a typo, because is says: note: suggested alternative: 'LinearizerCompo'

So, now I notice that you—for dreal4—you have specified a bespoke version of the ibex library build. Is this preferred/required, rather than the more recent ibex release (v.2.8.7)? It’s clear that this was not tried on RHEL7, because the waf switches “--with-solver” and “--lp-lib=clp” do not work there.

Any advice on how to proceed?  Thank you,

—Kent

**Comment by soonho-tri** [comment]
> Do I have to build ibex from “ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca”?

Yes.

> If so, then there is a problem, because waf does not recognize the specified options.

Which options?

> Is this preferred/required, rather than the more recent ibex release (v.2.8.7)?

Yes, that's the case.

**Comment by DKPeterson** [comment]
Same problem:

[kent@localhost dreal4]$ bazel clean --expunge
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
[kent@localhost dreal4]$ BAZEL_LINKLIBS=-l%:libstdc++.a bazel build //...
Starting local Bazel server and connecting to it...
... still trying to connect to local Bazel server after 10 seconds ...
INFO: Analyzed 520 targets (80 packages loaded, 3397 targets configured).
INFO: Found 520 targets...
ERROR: /home/kent/sally/packages/dreal4/dreal/contractor/BUILD.bazel:14:17: C++ compilation of rule '//dreal/contractor:contractor' failed (Exit 1) gcc failed: error executing command /opt/rh/devtoolset-7/root/usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG ... (remaining 92 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
In file included from dreal/contractor/contractor_cell.cc:8:0:
./dreal/contractor/contractor_ibex_polytope.h:63:25: error: 'LinearizerCombo' is not a member of 'ibex'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
./dreal/contractor/contractor_ibex_polytope.h:63:25: note: suggested alternative: 'LinearizerCompo'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
                         LinearizerCompo
./dreal/contractor/contractor_ibex_polytope.h:63:25: error: 'LinearizerCombo' is not a member of 'ibex'
./dreal/contractor/contractor_ibex_polytope.h:63:25: note: suggested alternative: 'LinearizerCompo'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
                         LinearizerCompo
./dreal/contractor/contractor_ibex_polytope.h:63:40: error: template argument 1 is invalid
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                                        ^
./dreal/contractor/contractor_ibex_polytope.h:63:40: error: template argument 2 is invalid
INFO: Elapsed time: 1096.558s, Critical Path: 74.26s
INFO: 119 processes: 119 linux-sandbox.
FAILED: Build did NOT complete successfully

But I suspect it’s still using the ibex static library that I built for all of my other solvers, which is in /usr/local/lib/ and has these modules:

[kent@localhost dreal4]$ nm  /usr/local/lib/libibex.a|grep LinearizerCom
ibex_LinearizerCompo.cpp.3.o:
0000000000000000 t _GLOBAL__sub_I__ZN4ibex15LinearizerCompoC2ERNS_10LinearizerES2_
0000000000000000 T _ZN4ibex15LinearizerCompo12add_propertyERKNS_14IntervalVectorERNS_13BoxPropertiesE
0000000000000110 T _ZN4ibex15LinearizerCompo9linearizeERKNS_14IntervalVectorERNS_8LPSolverE
0000000000000030 T _ZN4ibex15LinearizerCompo9linearizeERKNS_14IntervalVectorERNS_8LPSolverERNS_13BoxPropertiesE
00000000000000c0 T _ZN4ibex15LinearizerCompoC1ERNS_10LinearizerES2_
00000000000000c0 T _ZN4ibex15LinearizerCompoC2ERNS_10LinearizerES2_
00000000000000a0 T _ZN4ibex15LinearizerCompoD0Ev
0000000000000090 T _ZN4ibex15LinearizerCompoD1Ev
0000000000000090 T _ZN4ibex15LinearizerCompoD2Ev
0000000000000000 V _ZTIN4ibex15LinearizerCompoE
0000000000000000 V _ZTSN4ibex15LinearizerCompoE
0000000000000000 V _ZTVN4ibex15LinearizerCompoE

Whereas the bespoke ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca dynamic library that I built especially for dreal4 is in /opt/libibex/2.8.7/lib/ and has these modules:

[kent@localhost dreal4]$ nm -D /opt/libibex/2.8.7/lib/libibex.so|grep LinearizerCombo
00000000002a9770 T _ZN4ibex15LinearizerCombo12add_propertyERKNS_14IntervalVectorERNS_13BoxPropertiesE
00000000002a9940 T _ZN4ibex15LinearizerCombo9linearizeERKNS_14IntervalVectorERNS_8LPSolverE
00000000002a96d0 T _ZN4ibex15LinearizerCombo9linearizeERKNS_14IntervalVectorERNS_8LPSolverERNS_13BoxPropertiesE
00000000002a97c0 T _ZN4ibex15LinearizerComboC1ERKNS_6SystemENS0_11linear_modeE
00000000002a97c0 T _ZN4ibex15LinearizerComboC2ERKNS_6SystemENS0_11linear_modeE
00000000002a96b0 T _ZN4ibex15LinearizerComboD0Ev
00000000002a9670 T _ZN4ibex15LinearizerComboD1Ev
00000000002a9670 T _ZN4ibex15LinearizerComboD2Ev
000000000089b640 V _ZTIN4ibex15LinearizerComboE
00000000005f5b70 V _ZTSN4ibex15LinearizerComboE
000000000089b658 V _ZTVN4ibex15LinearizerComboE

As I showed, above, I’m using this bazel build command:

BAZEL_LINKLIBS=-l%:libstdc++.a bazel build //...

How do I direct bazel to use the special-purpose ibex library for building dreal, instead?

—Kent

**Comment by soonho-tri** [comment]
Have you tried this part of the doc? It's supposed to build ibex-2.7.4 and install at /opt/libibex/2.7.4. Then, dReal4 is supposed to pick it up from there.

```bash
curl https://codeload.github.com/dreal-deps/ibex-lib/tar.gz/fde1b111a5439eb59b7260bb84189e1a0c9cffca | tar xz
cd ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca/
CXXFLAGS=-std=c++0x ./waf configure \
	--enable-shared \
	--with-optim \
	--with-solver \
	--with-affine-extended \
	--interval-lib=filib \
	--lp-lib=clp \
	--prefix=/opt/libibex/2.7.4
sudo ./waf install
```

**Comment by DKPeterson** [comment]
Yes, I did try that. And this is what the logs look like. First the waf configure log:

[kent@localhost ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca]$ CXXFLAGS=-std=c++0x ./waf configure --enable-shared --with-optim --with-solver --with-affine-extended --interval-lib=filib --lp-lib=clp --prefix=/opt/libibex/2.8.7
Setting top to                           : /home/kent/sally/packages/ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca 
Setting out to                           : /home/kent/sally/packages/ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca/__build__ 
Checking for 'g++' (C++ compiler)        : /opt/rh/devtoolset-7/root/usr/bin/g++ 
Checking for 'gcc' (C compiler)          : /opt/rh/devtoolset-7/root/usr/bin/gcc 
Checking for program 'bison'             : /usr/bin/bison 
Checking for program 'flex'              : /usr/bin/flex 
.
.
.
plugin IbexSolve                                      : enabled 
Checking for compiler flags -std=c++11                : yes 
Checking for compiler flags -U__STRICT_ANSI__         : yes 
Configuration of the src directory 
Configuration of the tests 
Checking for 'cppunit'                                : not found 
Checking for header cppunit/Test.h                    : not found 
Checking for library cppunit                          : not found 
Load benchmarks module 
Checking for program 'gnuplot'                        : not found 
'configure' finished successfully (13m40.562s)

(Note that I accidentally used “--prefix=/opt/libibex/2.8.7” I could redo the whole build with “2.7.4” instead, but I think that only affects the metadata, not the contents. Right?)

And then the waf install log:

[kent@localhost ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca]$ sudo ./waf install
[sudo] password for kent: 
Waf: Entering directory `/home/kent/sally/packages/ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca/__build__'
[  1/206] Processing src/parser/parser.yc
[  2/206] Compiling src/arithmetic/ibex_Interval.cpp.in
[  3/206] Compiling src/numeric/ibex_LPSolver.cpp.in
[  4/206] Compiling src/symbolic/ibex_ExprOperators.cpp.in
[  5/206] Compiling src/parser/lexer.l
[  6/206] Compiling ibex.pc.in
.
.
.
[199/206] Linking __build__/src/libibex.so
[200/206] Compiling plugins/optim/main/ibexopt.cpp
[201/206] Compiling plugins/solver/main/ibexsolve.cpp
[202/206] Linking __build__/plugins/optim/ibexopt
[203/206] Linking __build__/plugins/solver/ibexsolve
+ install /opt/libibex/2.8.7/lib/libibex.so (from __build__/src/libibex.so)
+ install /opt/libibex/2.8.7/bin/ibexopt (from __build__/plugins/optim/ibexopt)
+ install /opt/libibex/2.8.7/bin/ibexsolve (from __build__/plugins/solver/ibexsolve)
Waf: Leaving directory `/home/kent/sally/packages/ibex-lib-fde1b111a5439eb59b7260bb84189e1a0c9cffca/__build__'
'install' finished successfully (23m34.916s)

And then, repeating what I sent earlier, the bazel build log:

bazel clean --expunge
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
[kent@localhost dreal4]$ BAZEL_LINKLIBS=-l%:libstdc++.a bazel build //...
Starting local Bazel server and connecting to it...
... still trying to connect to local Bazel server after 10 seconds ...
INFO: Analyzed 520 targets (80 packages loaded, 3397 targets configured).
INFO: Found 520 targets...
ERROR: /home/kent/sally/packages/dreal4/dreal/contractor/BUILD.bazel:14:17: C++ compilation of rule '//dreal/contractor:contractor' failed (Exit 1) gcc failed: error executing command /opt/rh/devtoolset-7/root/usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG ... (remaining 92 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
In file included from dreal/contractor/contractor_cell.cc:8:0:
./dreal/contractor/contractor_ibex_polytope.h:63:25: error: 'LinearizerCombo' is not a member of 'ibex'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
./dreal/contractor/contractor_ibex_polytope.h:63:25: note: suggested alternative: 'LinearizerCompo'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
                         LinearizerCompo
./dreal/contractor/contractor_ibex_polytope.h:63:25: error: 'LinearizerCombo' is not a member of 'ibex'
./dreal/contractor/contractor_ibex_polytope.h:63:25: note: suggested alternative: 'LinearizerCompo'
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                         ^~~~~~~~~~~~~~~
                         LinearizerCompo
./dreal/contractor/contractor_ibex_polytope.h:63:40: error: template argument 1 is invalid
   std::unique_ptr<ibex::LinearizerCombo> linear_relax_combo_;
                                        ^
./dreal/contractor/contractor_ibex_polytope.h:63:40: error: template argument 2 is invalid
INFO: Elapsed time: 1096.558s, Critical Path: 74.26s
INFO: 119 processes: 119 linux-sandbox.
FAILED: Build did NOT complete successfully

So, I think I’m up to speed with your recent  GitHub instructions: “How to Build and Install dReal on Fedora 32 <https://github.com/dreal/dreal4/blob/6f39e530024f536d26161a07fe0774caa544789d/docs/BuildOnFedora.md#how-to-build-and-install-dreal-on-fedora-32>”.

As I said earlier, I think that bazel is trying to link to the generic ibex library (that I use for other solvers), rather than the bespoke version you have spec’d. How do I force bazel to use the version in /opt/libibex/2.8.7/lib/libibex.so?

—Kent

**Comment by soonho-tri** [comment]
> (Note that I accidentally used “--prefix=/opt/libibex/2.8.7” I could redo the whole build with “2.7.4” instead, but I think that only affects the metadata, not the contents. Right?)

FYI, it needs to be `/opt/libibex/2.7.4` because of these lines, https://github.com/dreal/dreal4/blob/master/dreal/workspace.bzl#L25-L26.
